### DISCUSSION TITLE
fast options have very small real impact on speed?

### DISCUSSION THREAD

**Discussion by JM-Bader** [vdemichev/DiaNN#422] [Q&A]
I ran multiple variations of the same DIANN1.8.1 GUI search with the aim of testing how the speed can be enhanced. 

There were 96 raw files (in either .d or .d.dia format) placed on a local SATA (non-SSD) hard drive disk. I used a library. 
I extracted the times of 4 main steps of the search from the log files. Steps were: i) first pass search, ii) first cross-run analysis (improves search-specific library), iii) second pass search, iv) second cross run analysis (yielding final quantitative data). 

search times (on y-axis) in seconds based on various search settings (x axis - see code below)
![image](https://user-images.githubusercontent.com/98888664/176255080-5b1e6734-17a6-45a8-9e62-5344b1d43d59.png)
![image](https://user-images.githubusercontent.com/98888664/176255234-1db0810f-5339-4c7a-942f-8872b4586775.png)

1) .d.dia is generally faster than .d as a raw file format. 
2) the library-enhancing first cross run analysis is very time-consuming, when using the default smart profiling setting:
![image](https://user-images.githubusercontent.com/98888664/176256141-2265d762-c0b1-4c76-bca8-d4ccb6d8553b.png)

It can be disabled by the library generation setting IDs, RT & IM profiling ("IDRTIM"):
![image](https://user-images.githubusercontent.com/98888664/176255504-fdc51f3d-d620-47b8-b2a3-142ae767a899.png)

However, while this dramatically reduces the length of the first cross run analysis, it also surprisingly increases the length of the Second Pass search - which almost nullifies the time gained. (compare search #1 and #5)

>> Q1: Why is that the case? Can that be prevented?

3) Use of alternative Speed and RAM usage settings appear to make almost no difference on the search speed when using the fast .d.dia files: 

default setting: 
![image](https://user-images.githubusercontent.com/98888664/176256885-7c47948c-da94-449e-b374-ef27bcec4b76.png)
alternative settings: 
![image](https://user-images.githubusercontent.com/98888664/176256953-28f150e4-7684-4568-a0cf-ff65e4b2e6f4.png)

resulting search times: 
![image](https://user-images.githubusercontent.com/98888664/176256757-855d66ad-32ef-4818-9e3c-5701514588b3.png)
![image](https://user-images.githubusercontent.com/98888664/176256783-1f118be9-3874-41dd-a919-547c21003c3a.png)

>> Q2: Is that normal (that low RAM & high Speed or ultra-fast actually don't really result in a big speed difference)? Why is that the case? Can it be fixed?

>> Q3: Are there other options to make the search faster?

**Comment by vdemichev** [comment]
Hi, thanks for testing! I think alternative speed settings have little effect as even with .d.dia ~half of the time is being spend on loading those raw files from disk. Meaning it's not a search-intense dataset.

Q1. More fragments in the library => longer search. No, no need to prevent this, the effect is minor & good for IDs.
Q2. Yes, see above.
Q3. No, cannot speed up loading from disk.

Vadim

**Reply by JM-Bader** [reply]
Reg. answer to Q1: The second pass search is longer when using ID, RT & IM profiling compared to smart profiling. You said this is due to more fragments in the library which make the search longer. 

Q4: I would expect that the processing time per file is longer as there are more fragments to be searched. However, the processing time per raw file of the second pass search seems to be unchanged while the loading time gets longer: 
![image](https://user-images.githubusercontent.com/98888664/176483422-d53a59b0-1dab-499e-99bb-9fd9355e31c2.png)
Times are sums across all raw files. pcmann406 = Comp183 (same searches as data above)
 
Why is that? Does DIANN load more or less of the raw file depending on the number of fragments that it searches?
    
Q5: If ID, RT & IM profiling is good for IDs (with a minor effect), why is smart profiling the default (since it also appears to mildly slower) ?

**Reply by vdemichev** [reply]
Loading times probably depends on how you operating system is caching files to RAM. 
Smart profiling is a more universal method, likely better for PTMs for which fragmentation is not predicted well by the deep learning predictor.

**Reply by JM-Bader** [reply]
RE Q4: I see. But I guess that the search settings 'smart profiling' vs 'ID, RT & IM profiling' don't differ in terms of how my operating system is caching files to RAM. Or do they ?

**Reply by vdemichev** [reply]
The don't, probably random effect

**Comment by JM-Bader** [comment]
I also checked the proteome depth and consistency of results across search conditions, in particular for the fast 4 conditions #1, #5, #9, #13 (all .d.dia raw files, difference in smart profiling vs ID, RT & IM profiling // and optimal results vs lowRAM+fast vs ultrafast) - see top graph. 

The depth wasn't really different
However, biological differnence case/control (dataset = 96 samples, case and control) did not correlate as well I had expected on protein level: Pearson R = 0.86 (comparing case/control for all proteins in search #1 (library generation = smart profiling) vs search #3 (Library generation = ID, RT & IM profiling) with other settings identical. 

Q1: Do you think this is normal ? Isn't this low ?
Q2: Where does the difference come from ? 
- Different quantification of same fragments? (I guess shouldn't be impacted by library generation) 
- Or different fragments? (That many different ones? > such strong impact on quantification?)

Q3: Which one is likely more stringent/better for quantification setting (smart profiling vs ID, RT & IM profiliing)?
Q4: Is there a good way to assess that ? (Checking CVs for replicates ? other ideas?)

**Reply by vdemichev** [reply]
Q1: yes, probably all normal.
Q2: likely different fragments.
Q3: likely comparable.

**Reply by JM-Bader** [reply]
Re Q3: Well, if the searches differ so strongly (r = 0.86), isn't it likely that one of them is better than the other one (i.e. closer to the truth regarding the quantitation of proteins)? If I understand you correctly, you say that despite the differences both searches / settings are equally close/distant to the true quantities - correct?

**Reply by vdemichev** [reply]
Yes

**Comment by JM-Bader** [comment]
I noticed that setting Speed and RAM usage to 'Low RAM and high Speed' ("fast" in image below) actually leads to mildly higher proteome depth compared to 'optimal Results'

![image](https://user-images.githubusercontent.com/98888664/178686299-fe1ca938-ed31-48db-821f-0ed6938c9753.png)

Q1: Isn't that surprising?
Q2: Why is that?

**Reply by vdemichev** [reply]
Q1-2: a different algorithm, can indeed result in somewhat better results in some cases. There's no specific reason for this, just appears that this algorithm is better on this dataset. What is surprising, is very good performance of ultra-fast in this benchmark.
